### DISCUSSION TITLE
Next Js App Directory (Internationalized Routing i18n)

### DISCUSSION THREAD

**Discussion by AwaisTariq18** [vercel/next.js#45429] [Help]
I'm working on a project using  app directory for the first time and I'm confused and I need to implement a language switcher in which all the data with different language for pages will come from an API and I saw that we can use i18n for locale but in the app directory I cant seem to get the values of locale and in app directory docs there was nothing to help for this. Need Some Help

**Comment by guitheengineer** [comment]
Hey @AwaisTariq18, I recommend reading https://beta.nextjs.org/docs/guides/internationalization

**Comment by munkacsimark** [comment]
Here is the documentation related to `app` directory:
https://beta.nextjs.org/docs/guides/internationalization

**Comment by teobler** [comment]
I have some issues related to i18n as well, but example and docs could not help on my scenario.
example and docs suggest app/ dir looks like this:
|------------app
|--------[lang]
|----page.js
|----xxxx.js

that means your url will be something like:
https://xxx.com/[lang]/xxx

in my scenario, I want root url to be a default lang(eg. en), then other lang will use suffix path to point out, then it will be:
- https://xxx.com/ for en
- https://xxx.com/[lang]/ for others

then my app dir will be something like:
|------------app
|--------layout.js
|--------[lang]
|----layout.js
|----page.js
|----xxxx.js

then I will set `html` element `lang` attribute as `en`.
use `generateStaticParams` in [lang] folder layout.js to detect other `lang` for setup.

then I get a warning:
`Warning: You are mounting a new html component when a previous one has not first unmounted. It is an error to render more than one html component at a time and attributes and children of these components will likely fail in unpredictable ways. Please only render a single instance of <html> and if you need to mount a new one, ensure any previous ones have unmounted first.`

my question is, how can I modify lang attribute safety? Or if I'm doing wrong, what's the correct way to implement this scenario?

**Reply by teobler** [reply]
and if I access https://xxx.com/[lang]/ directly, I cannot change lang anymore, seems server will use root layout file.

**Reply by LoyallAmsterdam** [reply]
You can use [lang] for all pages and use a 308 redirect for all /default language paths?

Have you been able to setup translated slugs as well? 

en: /about-us
de: /de/uber-uns
nl: /nl/over-ons

**Reply by svobik7** [reply]
For translated slugs I would drop the [lang] approach and utilize static routes. I made a tiny lib for generating those translated static routes. Feel free to check it might help you.

https://github.com/svobik7/next-roots
